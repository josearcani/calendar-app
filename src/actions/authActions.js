import { types } from '../types/types';
import { fetchWithoutToken, fetchWithToken } from '../helpers/fetch';
import Swal from 'sweetalert2';

export const startLogin = (email, password) => {
  return async (dispatch) => {

    const resp = await fetchWithoutToken('auth', { email, password }, 'POST');
    const body = await resp.json(); // {name,ok,token,uid}
    
    if (body.ok) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime()); // to know when token was created

      dispatch(login({ name: body.name, uid: body.uid }));

    } else {
      Swal.fire('Error', body.msg, 'error');
    }

  }
}

export const startRegister = (name, email, password) => {
  return async (dispatch) => {

    const resp = await fetchWithoutToken('auth/new', {name, email, password }, 'POST');
    const body = await resp.json();

    // console.log('startRegister', body);
    if (body.ok) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(login({ name: body.name, uid: body.uid }));

    } else {
      Swal.fire('Error', body.msg, 'error');
    }

  }
}

export const startChecking = () => {
  return async (dispatch) => {

    const resp = await fetchWithToken('auth/renew');
    const body = await resp.json();

    // console.log('startChecking', body);
    if (body.ok) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(login({ name: body.name, uid: body.uid }));
      
    } else {
      dispatch(checkingFinish());
    }
  }
}

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(logout());
  }
}

const logout = () => ({
  type: types.authLogout
});

const login = (user) => ({
  type: types.authLogin,
  payload: user
});

const checkingFinish = () => ({
  type: types.authCheckingFinish
});
