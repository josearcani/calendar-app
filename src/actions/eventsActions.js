import { fetchWithToken } from '../helpers/fetch';
import { prepareEvents } from '../helpers/prepare-events';
import { types } from '../types/types';

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event
});

export const startEventAddNew = (event) => {
  return async (dispatch, getState) => {
    const { uid, name } = getState().auth;
    try {
      const resp = await fetchWithToken('events', event, 'POST');
      const body = await resp.json();

      if (body.ok) {
        event.id = body.event.id;
        event.user = { _id: uid, name };
        dispatch(eventAddNew(event));
      } 
    } catch (error) {
      console.log(error);
    }
  }
}

export const eventStartLoading = () => {
  return async (dispatch) => {
    try {
      const resp = await fetchWithToken('events');
      const body = await resp.json(); // {ok: bool, events: arr, total: num}
      const events = prepareEvents(body.events);
      dispatch(eventLoaded(events));
    } catch (error) {
      console.log(error);
    }
  }
}

const eventLoaded = (events) => ({
  type: types.eventLoaded,
  payload: events
})

const eventAddNew = (event) => ({
  type: types.eventAddNew,
  payload: event
});

export const eventClearActiveEvent = () => ({
  type: types.eventClearActiveEvent,
});

export const eventUpdated = (event) => ({
  type: types.eventUpdated,
  payload: event
});

export const eventDeleted = () => ({
  type: types.eventDeleted,
});
