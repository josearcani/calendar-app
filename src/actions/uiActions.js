import { types } from '../types/types';

export const openModal = () => ({
  type: types.openModal,
});

export const closeModal = () => ({
  type: types.uiCloseModal,
});
