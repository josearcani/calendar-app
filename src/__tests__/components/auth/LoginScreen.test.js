import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startLogin, startRegister } from '../../../actions/authActions';
import Swal from 'sweetalert2';

jest.mock('sweetalert2', () => ({
  fire: jest.fn()
}));

jest.mock('../../../actions/authActions', () => ({
  startLogin: jest.fn(),
  startRegister: jest.fn()
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};
const store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={ store }>
    <LoginScreen />
  </Provider>
)

describe('Test on <LoginScreen />', () => {

  beforeEach(() => {
    jest.clearAllMocks();
  })

  test('should render normally', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should run login dispatch', () => {
    wrapper.find('input[name="lEmail"]').simulate('change', {
      target: {
        name: 'lEmail',
        value: 'john@email.com'
      }
    });
    wrapper.find('input[name="lPassword"]').simulate('change', {
      target: {
        name: 'lPassword',
        value: 'secret'
      }
    });

    wrapper.find('form').at(0).simulate('submit', { preventDefault(){} });
    expect(startLogin).toHaveBeenCalledWith('john@email.com', 'secret');
  });

  test('should FAIL register when passwords do not match', () => {
    wrapper.find('input[name="rPassword2"]').simulate('change', {
      target: {
        name: 'rPassword2',
        value: '123456'
      }
    })
    wrapper.find('form').at(1).simulate('submit', { preventDefault(){} });
    expect(startRegister).not.toHaveBeenCalled();
    expect(Swal.fire).toHaveBeenCalledWith('Error', 'Las contraseÃ±as deben de ser iguales', 'error');
  });
  
  test('should register normally', () => {
    wrapper.find('input[name="rPassword"]').simulate('change', {
      target: {
        name: 'rPassword',
        value: '123456'
      }
    })
    wrapper.find('input[name="rPassword2"]').simulate('change', {
      target: {
        name: 'rPassword2',
        value: '123456'
      }
    })
    wrapper.find('form').at(1).simulate('submit', { preventDefault(){} });
    expect(startRegister).toHaveBeenCalledWith('nataly', 'nataly@test.com', '123456');
    expect(Swal.fire).not.toHaveBeenCalled();
  });
});
