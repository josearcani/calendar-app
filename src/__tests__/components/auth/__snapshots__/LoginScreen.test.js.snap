// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <LoginScreen /> should render normally 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value="joseph@test.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="secret"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rName"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value="nataly"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value="nataly@test.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="secret"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value="secret"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
